text/0000-parallel-model-hooks.md

```diff
@@ -0,0 +1,115 @@
+
+`transition-state` presently has a `resolveOneHandlerInfo` method that recurses through the `handler-info`s for a transition and calls the associated `resolve` method. Before this, we insert a `requestOneHandlerInfo` method that recurses through `handler-info`s, calling each one's `request` method, which now executes the model hooks, kicking off any network requests.
+
+We add a `requestIndex` property to `transition` that keeps track of how many route's model hooks we have already run (akin to how `resolveIndex` presently keeps track of how many routes we have resolved).
+
+#### Ember.js (emberjs/ember.js#12415)
+
+The default model hook uses `requestIndex` (instead of `resolveIndex`) to grab `modelPromise` (instead of `context`) from the `handler-info`.
+
+`Route#modelFor` returns a promise.
+
+#### Usage
+
+Routes that don't depend on another route's data likely don't need any changes, but their requests are sent in parallel:
+
+```javascript
+// app/routes/parent.js
+export default Ember.Route.extend({
+  model() {
+    return Ember.$.get('/api/foo');
+  }
+});
+
+// app/routes/parent/child.js
+export default Ember.Route.extend({
+  model(params) {
+    return Ember.$.get(`/api/foo/${params.id}`);
+  }
+});
+```
+
+Routes that depend on another route's data must consume that data as a promise:
+
+```javascript
+// app/routes/parent.js
+export default Ember.Route.extend({
+  model(params) {
+    return Ember.$.get(`/api/foo/${params.id}`);
+  }
+});
+
+// app/routes/parent/child.js
+export default Ember.Route.extend({
+  model() {
+    return this.modelFor('parent').then((parentModel) => {
+      return Ember.$.get(`/api/bar/${parentModel.get('baz')}`);
+    });
+  }
+});
+```
+
+Routes whose `afterModel` hook operates on the resolved model must now consume a promise:
+
+```javascript
+export default Ember.Route.extend({
+  model(params) {
+    return Ember.$.get(`/api/foo/${params.id}`);
+  },
+  afterModel(modelPromise, transition) {
+    return modelPromise.then((data) => {
+      if (data.abort) {
+        this.transitionTo('fiz');
+      }
+    }
+  }
+});
+```
+
+# Drawbacks
+
+- This is a major breaking change.
+- Having `Route#modelFor` return a promise may be confusing.
+
+# Alternatives
+
+- Tie this paradigm change to something opt-in like routable components where people will already have to change their behavior.
+- Maintain it behind a feature flag on the Canary branch which you can toggle on and off via config until we can land it.
+- Provide some built-in hooks so that we're able to reach in and change this behavior as an addon in a safer way.
+- The worlds most absurd monkeypatch–which would also be packaged as an addon.
+- Maintain a patch and build our own custom Ember.
+
+# Unresolved questions
+
+
+- Should `modelFor` always return a promise? Should we instead add a `modelPromiseFor` method?
```

Comment:
One of the things that we're considering as part of this process is enforced disallowal to mutate state that can cause trouble during these hooks by creating a special `this` object/context. If we run things in this special context while this is feature-flagged we'll be able to identify the types of things people want to do and provide APIs for that behavior.

This is also a first step toward optional removal of dynamic dependency injection (think, addon) which will allow for fully-resolving dependency graphs at build time for things like app partitioning without strict development time partitioning with engines. (Otherwise anything which _could_ be dynamically injected will have to be part of the primary payload.)